### DISCUSSION TITLE
Can the dreaded return = new block issue be fixed?

### DISCUSSION THREAD

**Discussion by tigersoul925** [vzhd1701/enex2notion#26] [Ideas]
Notion starts a new block on return. This works fine when your paragraph look like the sentence you're reading right now. It will look the same when imported.

But if you type like this:
And add another row
And a third row

Then after an empty row, a new paragraph is started. This style of text will look like trash imported into Notion.
Could you add an option to only start a new block on an empty row?

**Comment by vzhd1701** [comment]
I've introduced new `--condense-lines` option that will compress lines into paragraphs. Please check out the new version and tell how it works for you.

**Reply by tigersoul925** [reply]
Great work! Generally it works great, but there are examples where it fails for some unknown reason. I've attached an example note where it happens. Look at the attached image from Notion import. At the line there's an incorrect block split.

https://drive.google.com/drive/folders/108e18EKLdzwRo2PmpMrvhnxJd0ifLnys?usp=sharing

EDIT: The issue is the `</div><div><div/><div>` in the 4: paragraph before "FÖRDEL: fungerar garenterat...". This is not reflected in any way in the actual text of the evernote window. I don't know why Evernote has added the empty `<div></div>` tag.

**Reply by tigersoul925** [reply]
Also, if there are several empty rows, they are still interpreted as a single block split. I can't decide whether this should be considered correct behavior or not. An alternative could be interpreting empty row 2 and onward as empty blocks. I.e. 2 empty rows = 1 block split, 1 empty block. Not sure about this one.

EDIT: the more I think about it, multiple empty rows should probably be handled. It's a common way to divide documents in a crude way.

**Reply by tigersoul925** [reply]
Finally after going over _many_ notes, I'd suggest adding a second version of the condense option available:

`Whenever a new block is started due to an empty line, finish the last paragraph with a soft enter. This produces a larger divide between the paragraphs.
`

The reason for the 2nd option is that some notes just won't look good with the condense option, the spacing between paragraphs becomes too small and the document is hard to read. Examples are sheet music, songs with chords etc. In this case it's however still useful to keep the paragraphs as blocks. Without any condensing at all, the readability is just as bad again.

For being clear: this shouldn't replace the current version of condensing, there should be the option for the user to select between the two. Unless you find a clever way to realize when to use which, but I doubt that would be easy to do.

Examples of the problem and suggested 2nd alt method can be found in the subfolder of this share: https://drive.google.com/drive/folders/108e18EKLdzwRo2PmpMrvhnxJd0ifLnys?usp=sharing

**Reply by vzhd1701** [reply]
`--condense-lines-sparse` option added, please check out the new version and tell how it works for you.

**Reply by vzhd1701** [reply]
> EDIT: The issue is the `</div><div><div/><div>` in the 4: paragraph before "FÖRDEL: fungerar garenterat...". This is not reflected in any way in the actual text of the evernote window. I don't know why Evernote has added the empty `<div></div>` tag.

Evernote messes up the formatting sometimes. I've added a fix that will ignore invisible blocks in future.

**Reply by tigersoul925** [reply]
the new --condense-lines-sparse works good! And the invisible blocks thing has been fixed nicely.
However: I've found several other situations where Evernote formatting clearly is screwed up. I'm afraid this may turn into a never ending cat mouse chase. An example:

`</span><br/></span></div><div><br/></div><div>`

This only looks like a single empty row in Evernote and should as such only be converted into a block split, but we end up with an empty row + a block split. I managed to fix it in many cases with this regex: 

`(?<!<div>|<br\/>|[^>])<br\/>`

But I'm not 100% sure it won't break other places. It's all a bit random.

**Reply by vzhd1701** [reply]
I can just trim all text blocks to remove superfluous empty rows for `--condense-lines` and `--condense-lines-sparse` options. It's less intrusive than directly messing with xml and fits the expected logic in my opinion. What do you think?

**Reply by tigersoul925** [reply]
That could work! The exception I guess is when the user has intentionally left multiple empty rows in the text. E.g.:

```
This is a long text about something bla blab lablbal
blablabla
blablabla



This is an older version of the text that we want seperated from the current version above.
```

**Reply by vzhd1701** [reply]
That will be split into 2 paragraphs anyway with the current logic.

**Reply by tigersoul925** [reply]
> That will be split into 2 paragraphs anyway with the current logic.

Then it sounds like a potentially smart idea!

**Reply by vzhd1701** [reply]
Now it will trim paragraphs, please check out the new version.

**Reply by tigersoul925** [reply]
> Now it will trim paragraphs, please check out the new version.

The new version handled some examples but not all. An example that wasn't handled is:

```
om en järnkärna används.<br/></div><div>
Magnetisk koppling:
```
This views as a normal soft enter in evernote (no empty row), but converts to an empty row.

My pre-editing solution handles these, but it also messes with the xml. What I do is regex-removing this:
`(?<!<div>|<br\/>)<br\/>(?![^<])`

The problem is `<br/>` tags that are weirdly used and interpreted by Evernote.
